### ISSUE TITLE
NA OFI: HG_Init fails on OFI/PSM2 if Intel MPI is also using PSM2

### ISSUE THREAD

**Issue by carns** [mercury-hpc/mercury#200]
Observed using Intel MPI on the LCRC Bebop system; example scripts are here:

https://xgitlab.cels.anl.gov/sds/ssg/tree/master/tests/perf-regression/bebop

The mpirun command is using -IB to force MPI to use Infiniband transport instead of it's default transport in order to make the example benchmark work properly.  Otherwise HG_Init() fails after triggering this mr_mode mismatch error in libfabric:

https://github.com/ofiwg/libfabric/blob/master/prov/psm2/src/psmx2_domain.c#L293

This is not a problem for standalone Mercury programs that don't use MPI.

We need to either figure out how to make na_ofi and Intel MPI interoperable on PSM2, or else document this issue.

**Comment by soumagne** [comment]
Hmm sorry for the delay but I'm finally set up on bebop :)

I could not reproduce that problem... my test is not doing any communication over MPI though, I am just using MPI as a launcher. I guess I should now try to have MPI initialized first and see if that changes.

**Comment by soumagne** [comment]
Hmm no calling MPI_Init() for me did not change anything, it works fine and I'm using Intel MPI as well...

**Comment by carns** [comment]
Hunh, that's weird.  Anything different between the initialization steps in https://xgitlab.cels.anl.gov/sds/ssg/blob/master/tests/perf-regression/margo-p2p-latency.c and what you are testing?

**Comment by soumagne** [comment]
which intel MPI are you using? I'm using the default one but I'm wondering now if you are using another one?

**Comment by carns** [comment]
Default, compiling everything with either icc or mpiicc (mpicc uses GNU rather than Intel compilers by default).

**Comment by soumagne** [comment]
Hmm can you try again with the patches that are in now?

**Comment by liuxuezhao** [comment]
Hi, just FYI that some days ago we found similar problem that the MPI program (which called MPI_Init) cannot create HG class correctly (we were using public openmpi, not intel MPI).

The issue we met is by default the ompi will use the psm2 mtl and will open a psm2 port, this cause some conflictions with HG OFI usage. We workaround it by adding "--mca mtl ^psm2,ofi" options to orterun.

We did not use intel MPI, so maybe the problem is different.
To see why the HG_Init fails, can set ENV of "export FI_LOG_LEVEL=info" before loading the testing, it can report some OFI level log maybe helpful to see why it failed.

**Comment by carns** [comment]
FYI, I just tried with current origin/master and my test program still fails HG_Init() with psm2 if Intel MPI is using PSM2:

```
# NA -- Error -- /tmp/mochi-regression-168793/mercury/src/na/na_ofi.c:1191
 # na_ofi_domain_open(): fi_domain failed, rc: -22(Invalid argument).
# NA -- Error -- /tmp/mochi-regression-168793/mercury/src/na/na_ofi.c:1803
 # na_ofi_initialize(): Could not open domain for psm2, ib0
# NA -- Error -- /tmp/mochi-regression-168793/mercury/src/na/na.c:380
 # NA_Initialize_opt(): Could not initialize plugin
# HG -- Error -- /tmp/mochi-regression-168793/mercury/src/mercury_core.c:1039
 # hg_core_init(): Could not initialize NA class
# HG -- Error -- /tmp/mochi-regression-168793/mercury/src/mercury_core.c:3005
 # HG_Core_init_opt(): Cannot initialize HG core layer
# HG -- Error -- /tmp/mochi-regression-168793/mercury/src/mercury.c:905
 # HG_Init_opt(): Could not create HG class
Error: HG_Init()
```

As before, this goes away if I use -IB with mpirun in Intel MPI to force MPI itself to use Verbs.  I'll try some manual runs with debugging messages on to see if I can find out more.

**Comment by carns** [comment]
[psm2_with_mpi.txt](https://github.com/mercury-hpc/mercury/files/1647575/psm2_with_mpi.txt)
[psm2_with_mpi_ib.txt](https://github.com/mercury-hpc/mercury/files/1647576/psm2_with_mpi_ib.txt)

Two job output files attached, both with FI_LOG_LEVEL=1, and with the -l option on mpirun to prepend each line of stdout with the rank number that generated it.  The _ib.txt file is using the -IB mpirun option to force MPI to use verbs (which works) while the other is letting MPI try to use psm2 as well (which doesn't work).  This was all built with origin/master versions of psm2, libfabric, and mercury as of today.

The slurm job script looked like this:
```
#!/bin/bash
#SBATCH -N 2
#SBATCH -A radix
#SBATCH --ntasks-per-node=1
#SBATCH --time=15:00
#SBATCH -p bdwall

export FI_PSM2_NAME_SERVER=1
export FI_LOG_LEVEL=debug

echo "## Margo OFI/PSM2 (round trip, bdw):"
mpirun -l -genvall ./margo-p2p-latency -i 100000 -n psm2://ib0:5000
```

The _ib.txt was generated from the same script, just with -IB added to the mpirun command line.

This isn't a huge priority for me right this minute, just sharing data while I have the test environment handy.

**Comment by carns** [comment]
One other bit of detail on those logs.  Even though it is labeled as an "info" message in the output, I think that this is the key line that shows up on both ranks:

```
[1] libfabric:psm2:domain:psmx2_domain_open():294<info> mr_mode mismatch: expecting FI_MR_SCALABLE
```

This corresponds to this check in libfabric, where an error is returned from domain_open():

https://github.com/ofiwg/libfabric/blob/master/prov/psm2/src/psmx2_domain.c#L295

**Comment by soumagne** [comment]
Hmm it is weird, we do have a global domain list, which means that if you were initializing once the psm2 domain, it would be picked up again next time you call HG/NA_Init() but it would be initialized with the same mr mode...

**Comment by liuxuezhao** [comment]
I guess the problem is very similar as what we ever met in openmpi.
In openmpi's MPI_Init(), it possibly will create a fi_domain with FI_MR_SCALABLE mode, then later when HG OFI plugin create another one with FI_MR_BASIC mode the libfabric failed it.

In our testing, we workaround it by adding "--mca mtl ^psm2,ofi" options to orterun, by that the MPI_Init will not interact with psm2/ofi mtl and no fi_domain created and hence avoid the conflict.

It possibly is not a good enough solution as the user program cannot use OPA for MPI communication (but the HG level RPC communication is over OPA). Two refining approaches might be: 1) change OFI plugin to also work in FI_SCALABLE mode, but I guess this possibly also will cause some other issues for both MPI usage and HG RPC usage on the fi_domain (same or different domain); 2) discuss with PSM2 person to see if it  can be refined within libfabric PSM2 provider, but not sure if it will need to involve libpsm2 side in that case will be more difficult. (note that OFI PSM2 provider and libpsm2 are two different modules and maintained by two different teams).

**Comment by carns** [comment]
FYI, I tried a quick hack to Mercury to make na_ofi use MR_SCALABLE mode, which can be found in this branch: https://github.com/carns/mercury/tree/carns/dev-ofi-psm2-mr-scalable

That didn't resolve the problem, though.  If I allow MPI to use OPA, then the FI_LOG_LEVEL=debug shows this message before Mercury fails to initialize now:

```
libfabric:psm2:core:psmx2_trx_ctxt_alloc():315<warn> psm2_ep_open returns 12, errno=0
```

FWIW, the MR_SCALABLE hack does run correctly if I set ```export I_MPI_FABRICS=shm:ofa``` to make MPI us OFA intead of OPA.

Any other ideas?
